scripts/src/java/org/oppia/android/scripts/kdoc/KDocCheck.kt

```diff
@@ -0,0 +1,284 @@
+package org.oppia.android.scripts.kdoc
+
+import org.jetbrains.kotlin.cli.common.CLIConfigurationKeys
+import org.jetbrains.kotlin.cli.common.messages.MessageCollector
+import org.jetbrains.kotlin.cli.jvm.compiler.EnvironmentConfigFiles
+import org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment
+import org.jetbrains.kotlin.com.intellij.openapi.util.Disposer
+import org.jetbrains.kotlin.com.intellij.psi.PsiManager
+import org.jetbrains.kotlin.com.intellij.testFramework.LightVirtualFile
+import org.jetbrains.kotlin.config.CompilerConfiguration
+import org.jetbrains.kotlin.idea.KotlinFileType
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.KtClass
+import org.jetbrains.kotlin.psi.KtConstructorDelegationReferenceExpression
+import org.jetbrains.kotlin.psi.KtDeclaration
+import org.jetbrains.kotlin.psi.KtElement
+import org.jetbrains.kotlin.psi.KtFile
+import org.jetbrains.kotlin.psi.KtNamedFunction
+import org.jetbrains.kotlin.psi.KtObjectDeclaration
+import org.jetbrains.kotlin.psi.KtSecondaryConstructor
+import org.jetbrains.kotlin.psi.KtVariableDeclaration
+import org.jetbrains.kotlin.psi.doNotAnalyze
+import org.oppia.android.scripts.common.RepositoryFile
+import org.oppia.android.scripts.proto.KDocExemptions
+import java.io.File
+import java.io.FileInputStream
+import java.util.Scanner
+
+/**
+ * Script for ensuring that KDocs are present on all non-private:
+ * - Classes
+ * - Functions
+ * - Values/fields
+ * - Explicit constructors
+ * - Companion objects
+ * - Nested classes
+ * - Enums
+ * - Annotations
+ * - Interfaces
+ *
+ * Note: If any of the above member has an override modifier present, then it is automatically
+ * exempted for the KDoc check.
+ *
+ * Usage:
+ *   bazel run //scripts:kdoc_check -- <path_to_directory_root>
+ *
+ * Arguments:
+ * - path_to_directory_root: directory path to the root of the Oppia Android repository.
+ *
+ * Example:
+ *   bazel run //scripts:kdoc_check -- $(pwd)
+ */
+fun main(vararg args: String) {
+  // Path of the repo to be analyzed.
+  val repoPath = "${args[0]}/"
+
+  val kdocExemptiontextProto = "scripts/assets/kdoc_exemptions"
```

Comment:
One scenario I think we missed in past checks that is especially important for this one is outputting an error when the exemption list contains something no longer relevant. I think we actually want to make sure that, in general, exemption lists are always the complete picture on the status of the codebase.

Reply:
Added.